api/error_unexpected_response_test.go

```diff
@@ -0,0 +1,268 @@
+	// correctly. This ensures that the routing rule for this path is at least
+	// correct and that the testServer is passing its address to the client
+	// properly.
+	t.Run("ValidateServer", func(t *testing.T) {
+		n, _, err := c.Namespaces().Info("default", nil)
+		must.NoError(t, err)
+		var ns api.Namespace
+		err = unmock(t, mockNamespaceBody, &ns)
+		must.NoError(t, err)
+		must.Eq(t, ns, *n)
+	})
+
+	// WrongStatus tests that an UnexpectedResponseError is generated and filled
+	// with the correct data when a response code that the API client wasn't
+	// looking for is returned by the server.
+	t.Run("WrongStatus", func(t *testing.T) {
+		t.Parallel()
+		n, _, err := c.Namespaces().Info("badStatus", nil)
+		must.Nil(t, n)
+		must.Error(t, err)
+		t.Logf("err: %v", err)
+
+		ure, ok := err.(api.UnexpectedResponseError)
+		must.True(t, ok)
+
+		must.True(t, ure.HasStatusCode())
+		must.Eq(t, http.StatusAccepted, ure.StatusCode())
+
+		must.True(t, ure.HasStatusText())
+		must.Eq(t, http.StatusText(http.StatusAccepted), ure.StatusText())
+
+		must.True(t, ure.HasBody())
+		must.Eq(t, mockNamespaceBody, ure.Body())
+	})
+
+	// NotFound tests that an UnexpectedResponseError is generated and filled
+	// with the correct data when a `404 Not Found`` is returned to the API
+	// client, since the requireOK wrapper doesn't "expect" 404s.
+	t.Run("NotFound", func(t *testing.T) {
+		t.Parallel()
+		n, _, err := c.Namespaces().Info("wat", nil)
+		must.Nil(t, n)
+		must.Error(t, err)
+		t.Logf("err: %v", err)
+
+		ure, ok := err.(api.UnexpectedResponseError)
+		must.True(t, ok)
+
+		must.True(t, ure.HasStatusCode())
+		must.Eq(t, http.StatusNotFound, ure.StatusCode())
+
+		must.True(t, ure.HasStatusText())
+		must.Eq(t, http.StatusText(http.StatusNotFound), ure.StatusText())
+
+		must.True(t, ure.HasBody())
+		must.Eq(t, "Namespace not found", ure.Body())
+	})
+
+	// EarlyClose tests what happens when an error occurs during the building of
+	// the UnexpectedResponseError using FromHTTPRequest.
+	t.Run("EarlyClose", func(t *testing.T) {
+		t.Parallel()
+		n, _, err := c.Namespaces().Info("earlyClose", nil)
+		must.Nil(t, n)
+		must.Error(t, err)
+
+		t.Logf("e: %v\n", err)
+		ure, ok := err.(api.UnexpectedResponseError)
+		must.True(t, ok)
+
+		must.True(t, ure.HasStatusCode())
+		must.Eq(t, http.StatusInternalServerError, ure.StatusCode())
+
+		must.True(t, ure.HasStatusText())
+		must.Eq(t, http.StatusText(http.StatusInternalServerError), ure.StatusText())
+
+		must.True(t, ure.HasAdditional())
+		must.ErrorContains(t, err, "the body might be truncated")
+
+		must.True(t, ure.HasBody())
+		must.Eq(t, "{", ure.Body()) // The body is truncated to the first byte
+	})
+}
+
+// testServer creates a httptest.Server that can be used to serve simple mock
+// data, which is faster than starting a real Nomad agent.
+func testServer(t *testing.T) string {
+	grabber := portal.New(t)
+	ports := grabber.Grab(1)
```

Comment:
By default this will use the same `127.0.0.1` address as the `portal.Grabber` being used by `testutil.TestServer`. We should either configure this one to use a different address, or reactor both callers to use a common `Grabber` like we do for the non-api packages in https://github.com/hashicorp/nomad/blob/main/ci/ports.go#L15-L20